pkg/gce-cloud-provider/compute/gce-compute.go

```diff
@@ -1285,6 +1364,164 @@ func (cloud *CloudProvider) waitForSnapshotCreation(ctx context.Context, project
+		}),
+	}
+}
+
+// getFilteredTagsMap returns the map of tag keys and the tag values to apply on the resources after
+// filtering the tags already existing on a given resource.
+func getFilteredTagsMap(ctx context.Context, client *rscmgr.TagBindingsClient, parent string, tagsMap map[string]string) map[string]string {
+	if len(tagsMap) <= 0 {
+		klog.V(1).Infof("getFilteredTagsMap: tags map is empty")
+		return nil
+	}
+
+	filteredTagsMap := make(map[string]string, len(tagsMap))
+	for key, value := range tagsMap {
+		filteredTagsMap[key] = value
+	}
+
+	listBindingsReq := &rscmgrpb.ListEffectiveTagsRequest{
+		Parent: parent,
+	}
+	bindings := client.ListEffectiveTags(ctx, listBindingsReq)
+	for i := 0; i < 50; i++ {
+		binding, err := bindings.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil || binding == nil {
+			klog.Errorf("failed to list effective tags on %s compute resource: %v: %v", parent, binding, err)
+			break
+		}
+		namespacedTagKey := binding.GetNamespacedTagKey()
+		if _, exists := filteredTagsMap[namespacedTagKey]; exists {
+			delete(filteredTagsMap, namespacedTagKey)
+			klog.V(1).Infof("getFilteredTagsMap: skipping tag %s already exists on the %s compute resource", namespacedTagKey, parent)
+		}
+	}
+
+	return filteredTagsMap
+}
+
+// attachTagsToResource attaches tags to a compute resource. As GCP has a rate limit of 600
+// requests per minute, the tags list is filtered so that only the tags which are not attached
+// to the resource are attached. The calls to create the tag binding resource for the tags are
+// rate limited to 8 requests per second.
+func (cloud *CloudProvider) attachTagsToResource(
+	ctx context.Context,
+	tagsMap map[string]string,
+	project string,
+	resourceID uint64,
+	resourceType ResourceType,
+	location string,
+	isZonal bool,
+	resourceManagerHostSubPath string) error {
+	if len(tagsMap) <= 0 {
+		return nil
+	}
+
+	tagBindingsClient, err := createTagBindingsClient(ctx, cloud.tokenSource, location, resourceManagerHostSubPath)
+	if err != nil || tagBindingsClient == nil {
+		return fmt.Errorf("failed to create tag binding client for adding tags to %d compute %s: %w", resourceID, resourceType, err)
+	}
+	defer tagBindingsClient.Close()
+
+	var fullResourceID string
+	if location != "" {
+		if isZonal {
+			fullResourceID = fmt.Sprintf(zonalOrRegionalComputeParentPathFmt, project, "zones", location, resourceType, resourceID)
+		} else {
+			fullResourceID = fmt.Sprintf(zonalOrRegionalComputeParentPathFmt, project, "regions", location, resourceType, resourceID)
+		}
+	} else {
+		fullResourceID = fmt.Sprintf(globalComputeParentPathFmt, project, resourceType, resourceID)
+	}
+
+	filteredTagsMap := getFilteredTagsMap(ctx, tagBindingsClient, fullResourceID, tagsMap)
+
+	// GCP has a rate limit of 600 requests per minute, restricting
+	// here to 8 requests per second.
+	limiter := common.NewLimiter(gcpTagsRequestRateLimit, gcpTagsRequestTokenBucketSize, true)
```

Comment:
We are proactively restricting the number of tag binding requests for a resource to 8 req/sec and also we are using [retries](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/1377/files#diff-0a6cd0afae8c035b921beb3150d39b6955f7b07eedec752c8f06f669619bde57R1398-R1409) with exponential backoff when we hit the limit.
[Quota](https://cloud.google.com/resource-manager/docs/limits) that could get exhausted because of tag operations is specific to resource manager APIs, which shouldn't affect other resources created by the driver.

Driver creates Disk, Snapshot, Image resources but of these APIs, API for creating Disk alone support passing tags during creation, for other two resources we have to bind tags after the resource creation.

@mattcary Could you please help with your suggestions.

Reply:
I think one optimization would be to use a shared rate limiter, scoped to a CloudProvider client. This would allow each PDCSI controller instance to rate limit calls across CreateSnapshot/CreateVolume.

The 90 second initial delay seems a bit long. I think in the worst case you'd only need to wait for a maximum of 60 seconds until the per-minute quota is refilled (per the documentation: https://cloud.google.com/compute/api-quota#api-rate-limits).

Also, I don't think having an exponential backoff is that useful here (especially at a rate of 2.0), since the per-minute quota is a fixed window rate limit. You'll either be competing with another client, and exhaust quota before the next minute, or quota will refill and the bucket will be replenished. The only thing you may want to adjust is a negotiation scheme for the per-second bucket rate, but without direct knowledge of the request rate of any other client, this seems complicated to determine.

Reply:
> I think one optimization would be to use a shared rate limiter, scoped to a CloudProvider client. This would allow each PDCSI controller instance to rate limit calls across CreateSnapshot/CreateVolume.

Yeah, would it be required to scope to a CloudProvider client, since tag operations has it's own client and the rate limiter could be initiated as part of `init()` and used as a global instance just for the tags. Please let us know WDYT.

> The 90 second initial delay seems a bit long. I think in the worst case you'd only need to wait for a maximum of 60 seconds until the per-minute quota is refilled (per the documentation: https://cloud.google.com/compute/api-quota#api-rate-limits).

Yeah, initially while trying to understand the APIs and the quotas I had checked with 60 seconds initial delay, with 9 clients in parallel adding 50 tags to a resource each, which was creating a traffic of 16qps and 60 seconds reset wasn't working as expected with few tags attach operations taking more than 3 minutes. And also there is a limit on per second and per minute as per the doc [here](https://cloud.google.com/resource-manager/docs/limits)(I am not sure if both are considered). And I tried with 90s delay and had observed all the tag operations getting completed within 2mins, that was the reason to use the same. Since this also depends on factors like network latency, the observation may not be correct too. Please let us know if 90s would be okay or we can change it to 60s.

> Also, I don't think having an exponential backoff is that useful here (especially at a rate of 2.0), since the per-minute quota is a fixed window rate limit. You'll either be competing with another client, and exhaust quota before the next minute, or quota will refill and the bucket will be replenished. The only thing you may want to adjust is a negotiation scheme for the per-second bucket rate, but without direct knowledge of the request rate of any other client, this seems complicated to determine.

During the tests, I had checked for 1350 tag write requests which got completed in  ~6minutes, with the same backoff numbers used in code without quota exceeded errors. I think retry on quota exhaustion error is helping to an extent. Please let us know WDYT.

Reply:
@pwschuurman ^^